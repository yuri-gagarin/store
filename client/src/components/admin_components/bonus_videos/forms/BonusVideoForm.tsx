import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
}

interface Props {
  description: string;
  youTubeURL: string;
  vimeoURL: string;
  handleCreateProduct(data: FormState): void;
  handleUpdateProduct(data: FormState): void;
}


const ProductForm: React.FC<Props> = ({ description, youTubeURL, vimeoURL, handleCreateProduct, handleUpdateProduct }): JSX.Element => {
  
  const [ newForm, setNewForm ] = useState<boolean>(true)
  const [ formState, setFormState ] = useState<FormState>({ description, youTubeURL, vimeoURL });
  const bonusVideoFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  
  const handleYouTubeURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      youTubeURL: e.target.value
    });
  };
  const handleVimeoURLChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      vimeoURL: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateProduct(formState);
    } else {
      handleUpdateProduct(formState);
    }
  };

  useEffect(() => {
    if (description || youTubeURL || vimeoURL) {
      setNewForm(false);
    } else {
      setNewForm(true);
    }
  },  [description, youTubeURL, vimeoURL]);

  useEffect(() => {
    if (bonusVideoFormRef.current) {
      const elem = bonusVideoFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [bonusVideoFormRef]);
  
  return (
    <div className="createProductFormHolder" ref={bonusVideoFormRef}>
      <Form id="createProductForm">
        <Form.Field>
          <label>Video YouTube URL:</label>
          <input 
            onChange={handleYouTubeURLChange} 
            placeholder="YouTube URL here ..." 
            value={formState.youTubeURL}
          />
        </Form.Field>
        <Form.Field>
          <label>Video Vimeo URL:</label>
          <input 
            onChange={handleVimeoURLChange} 
            placeholder="Vimeo URL here..." 
            value={formState.vimeoURL}
          />
        </Form.Field>
        <Form.Field
          id='form-textarea-control-opinion'
          control={TextArea}
          label='Video Description'
          onChange={hadnleDescriptionChange}
          placeholder='Video description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' onClick={handleSubmit} content= "Save New Bonus Video" />
            : <Button type='submit' onClick={handleSubmit} content= "Update Bonus Video" />
         }  
      </Form>
    </div>
  );
};

export default ProductForm;