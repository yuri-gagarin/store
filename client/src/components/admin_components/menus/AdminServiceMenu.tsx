import React, { useState, useEffect, useRef } from "react";
import { Menu, MenuItemProps } from "semantic-ui-react";
// routing //
import { withRouter, RouteComponentProps, useRouteMatch } from "react-router-dom"
// css imports //
import "./css/adminServiceMenu.css";
import { AppAction } from "../../../state/Store";

interface Props extends RouteComponentProps {
  dispatch: React.Dispatch<AppAction>
}
const AdminServiceMenu: React.FC<Props> = ({ history, dispatch }): JSX.Element => {
  const [ scrolled, setScrolled ] = useState<boolean>(false);
  const [ activeItem, setActiveItem ] = useState<string>("view_all");
  const [ menuOpen, setMenuOpen ] = useState<boolean>(false);

  const match = useRouteMatch("/admin/home/my_services");
  const adminServiceMenuRef = useRef<HTMLDivElement>(document.createElement("div"));

  const handleItemClick = (e: React.MouseEvent, { name }: MenuItemProps): void => {
    setActiveItem(String(name));


    switch (name) {
      case "view_all": 
        history.push(match?.path + "/all");
        break;
      case "create":
        history.push(match?.path + "/create");
        break;
      case "manage":
        history.push(match?.path + "/manage");
        break;
      default: history.push("/admin/home");
    }
  }
 
  const scrollListener = () => {
    if (window.scrollY > 1) {
      setScrolled(true);
    } else if (window.scrollY === 0) {
      setScrolled(false)
    } 
  };

  useEffect(() => {
    setTimeout(() => {
      setMenuOpen(true);
    }, 200);
  }, []);

  useEffect(() => {

    window.addEventListener("scroll", scrollListener, true);
    return () => window.removeEventListener("scroll", scrollListener);

  }, [adminServiceMenuRef]);

  return (
    <div className={ scrolled ? "adminServiceMenuFixed menuScrolled" : "adminServiceMenuFixed"} ref={adminServiceMenuRef}>
      <Menu tabular className={ menuOpen ? "adminServiceMenu serviceMenuOpen" : "adminServiceMenu" }>
      <Menu.Item
          name='view_all'
          content="View All Services"
          active={activeItem === 'view_all'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='create'
          content="Create New Service"
          active={activeItem === 'create'}
          onClick={handleItemClick}
        />
        <Menu.Item
          name='manage'
          content="Manage Services"
          active={activeItem === 'manage'}
          onClick={handleItemClick}
        />
      </Menu>
    </div>
  )
};

export default withRouter(AdminServiceMenu);