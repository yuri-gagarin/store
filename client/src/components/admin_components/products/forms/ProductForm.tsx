import React, { useState, useEffect, useRef, memo }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";
// type definitions//
import { ProductFormState } from "../type_definitions/productTypes";

interface Props {
  name: string;
  price: string;
  description: string;
  details: string;
  newForm: boolean;
  handleCreateProduct(data: ProductFormState): void;
  handleUpdateProduct(data: ProductFormState): void;
}
class ProductForm extends React.Component<Props, ProductFormState> {
  constructor(props: Props) {
    super(props)
    // this.productFormRef = useRef<HTMLDivElement>(document.createElement("div"));
    this.state = {
      name: "",
      price: "",
      description: "",
      details: "",
    }

  }
  // local form state //
  // form ref //
  
  // form state changes listeners //
  handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    this.setState({
      ...this.state,
      name: e.target.value
    });
  };
  handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      price: e.target.value
    });
  };
  hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      description: e.target.value
    });
  };
  handleDetailsChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({
      ...this.state,
      details: e.target.value
    });
  };
  // submit action //
  handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (this.props.newForm) {
      this.props.handleCreateProduct(this.state);
    } else {
      this.props.handleUpdateProduct(this.state);
    }
  };
  componentDidMount() {

  }
  shouldComponentUpdate(nextProps: Props, nextState: ProductFormState) {
    console.log(nextProps)
    const { name: newName } = nextProps;
    if (newName === this.props.name) {
      console.log("shouldnt");
      return false
    }
    return true;
  }

  // lifecycle hooks //
  /*
  useEffect(() => {
    if (productFormRef.current) {
      const elem = productFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [productFormRef]);
  */
  render() {
    const { name, price, description, details } = this.state;
    const { newForm } = this.props;
    return (
      <div className="productFormDiv">
        <Form id="productForm">
          <Form.Field>
            <label>Product name</label>
            <input 
              id="adminProductFormNameInput"
              onChange={this.handleNameChange} 
              placeholder="Product name here ..." 
              value={name}
            />
          </Form.Field>
          <Form.Field>
            <label>Product price</label>
            <input 
              id="adminProductFormPriceInput"
              onChange={this.handlePriceChange} 
              placeholder="Product price here..." 
              value={price}
            />
          </Form.Field>
          <Form.Field
            id='adminProductFormDescInput'
            control={TextArea}
            label='Product Description'
            onChange={this.hadnleDescriptionChange}
            placeholder='Product description here...'
            value={description}
           />
          <Form.Field
            id='adminProductFormDetailsInput'
            control={TextArea}
            label='Product Details'
            onChange={this.handleDetailsChange}
            placeholder='Product details here...'
            value={details}
          />
        </Form>
        {
          newForm 
            ? <Button 
                id="adminProductFormCreate" 
                type="button"
                onClick={this.handleSubmit} 
                content= "Create  New Product" 
              />
            : <Button 
              id="adminProductFormUpdate" 
              type='button'
              onClick={this.handleSubmit} 
              content="Update Product" 
            />
        }  
      </div>
    );
  }
}

export default ProductForm;