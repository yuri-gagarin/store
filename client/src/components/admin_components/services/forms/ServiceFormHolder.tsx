import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// client routing //
import { withRouter, RouteComponentProps } from "react-router-dom";
// css imports //
import "./css/serviceFormHolder.css";
// additional components //
import ServiceForm from "./ServiceForm";
import ServiceImgPreviewHolder from "../image_preview/ServiceImgPreviewHolder";
import ServiceImgUplForm from "./ServiceImgUplForm";
import LoadingBar from "../../miscelaneous/LoadingBar";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createService, editService } from "../actions/APIServiceActions";
import { openServiceForm, closeServiceForm } from "../actions/UIServiceActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { FormState } from "./ServiceForm";
interface Props extends RouteComponentProps {
  
}

type ServiceData = {
  name: string;
  description: string;
  price: string;
  images: IServiceImgData[];
}

const ServiceFormHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentServiceData, serviceFormOpen, error } = state.serviceState;
  const { _id : serviceId, name, description, price, createdAt, editedAt } = currentServiceData;
  // local component state //
  const [ newForm, setNewForm ] = useState<boolean>(true);

  // ServiceForm toggle //
  const toggleServiceForm = () => {
    serviceFormOpen ? closeServiceForm(dispatch) : openServiceForm(dispatch);
  };
  // API handlers CREATE - EDIT //
  const handleCreateService = ({ name, price, description }: FormState): void => {
    const serviceData: ServiceData = {
      name,
      description,
      price,
      images: currentServiceData.images
    };

    createService(serviceData, dispatch)
      .then((_) => {
        // service created //
        closeServiceForm(dispatch);
      })
      .catch((_) => {
        // handle an error here possibly show details //
      });
  };
  const handleUpdateService = ({ name, description, price }: FormState): void => {
    const serviceParams: ServiceData = {
      name, description, price, images: currentServiceData.images
    };

    editService(currentServiceData._id, serviceParams, dispatch, state)
      .then((_) => {
        closeServiceForm(dispatch);
      })
      .catch((_) => {
        // handle an error , maybe show modal in thhe future //
      });
  };
  // lifecyle hooks //
  useEffect(() => {
    if (!serviceFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ serviceFormOpen ]);
  // set correct data if a service is loaded //
  useEffect(() => {
      checkSetValues(currentServiceData) ? setNewForm(false) : setNewForm(true);
  }, [ currentServiceData ]);
  // component return //
  return (
    <div id="serviceFormHolder">
      { loading ? <LoadingBar /> : null }
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row id="serviceFormHolderDetails">
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="serviceFormHolderDetailsName">
                  <h3>Service name:</h3>
                  <p>{name}</p>
                </div>
                <div className="serviceFormHolderDetailsPrice">
                  <h3>Service price:</h3>
                  <p>{price}</p>
                </div>
                <div className="serviceFormHolderDetailsDesc">
                  <h3>Service description:</h3>
                  <p>{description}</p>
                </div>
                <div className="serviceFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <ServiceImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <ServiceImgUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button 
            id="serviceFormToggleBtn" 
            onClick={toggleServiceForm} 
            content={ !serviceFormOpen ? "Open Form" : "Close Form"}
          />
          {
            serviceFormOpen ? 
              <ServiceForm 
                name={name} 
                description={description} 
                price={price}
                newForm={newForm}
                handleCreateService={handleCreateService}
                handleUpdateService={handleUpdateService}
              /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

// export for testing withour router //
export { ServiceFormHolder };
// default export //
export default withRouter(ServiceFormHolder);

