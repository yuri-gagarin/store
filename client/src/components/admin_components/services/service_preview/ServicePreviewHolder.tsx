import React, { useEffect } from "react";
import { Grid, Item } from "semantic-ui-react";
// additional components //
import ServicePreview from "./ServicePreview";
import ServicesDetails from "./ServicesDetails"
// types and interfaces //
import { AppAction, IGlobalAppState } from "../../../../state/Store";
// api actions //
import { getAllServices } from "../actions/APIServiceActions";
import PopularServiceHolder from "./popular_services/PopularServiceHolder";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const ServicePreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedServices } = state.serviceState;
  console.log(loadedServices);
  useEffect(() => {
    getAllServices(dispatch);
  }, []);

  return (
    <Grid stackable padded columns={2}>
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedServices.map((service) => {
              return (
                <ServicePreview 
                  key={service._id}
                  service={service}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <ServicesDetails totalServices={loadedServices.length} />
        <PopularServiceHolder popularServices={loadedServices}/>
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
  )
};

export default ServicePreviewHolder;