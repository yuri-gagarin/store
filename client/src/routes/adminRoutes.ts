export const AdminProductRoutes = {
  HOME_ROUTE: "/admin/home/my_products",
  VIEW_ALL_ROUTE: "/admin/home/my_products/view_all",
  CREATE_ROUTE: "/admin/home/my_products/create",
  MANAGE_ROUTE: "/admin/home/my_products/manage",
  SHOW_ROUTE: "/admin/home/my_products/show",
  VIEW_SORTED_ROUTE: "/admin/home/my_products/view_sorted"
};

export const AdminServiceRoutes = {
  ADMIN_SERVICES_HOME_ROUTE: "/admin/home/my_services",
  ADMIN_SERVICES_VIEW_ALL_ROUTE: "/admin/home/my_services/view_all",
  ADMIN_SERVICES_CREATE_ROUTE: "/admin/home/my_services/create",
  ADMIN_SERVICES_MANAGE_ROUTE: "/admin/home/my_services/manage",
  ADMIN_SERVICES_SHOW_ROUTE: "/admin/home/my_services/show",
  ADMIN_SERVICES_VIEW_SORTED_ROUTE: "/admin/home/my_services/view_sorted"
};

export const AdminStoreRoutes = {
  ADMIN_STORES_HOME_ROUTE: "/admin/home/my_stores",
  ADMIN_STORES_CREATE_ROUTE: "/admin/home/my_stores/create",
  ADMIN_STORES_MANAGE_ROUTE: "/admin/home/my_stores/manage",
  ADMIN_STORES_SHOW_ROUTE: "/admin/home/my_stores/show",
  ADMIN_STORES_VIEW_ALL_ROUTE: "/admin/home/my_stores/view_all"
};

export const AdminStoreItemRoutes = {
  HOME_ROUTE: "/admin/home/my_store_items",
  CREATE_ROUTE: "/admin/home/my_store_items/create",
  MANAGE_ROUTE: "/admin/home/my_store_items/manage",
  SHOW_ROUTE: "/admin/home/my_store_items/show",
  VIEW_ALL_ROUTE: "/admin/home/my_store_items/view_all",
  VIEW_SORTED_ROUTE: "/admin/home/my_store_items/view_sorted"
};

