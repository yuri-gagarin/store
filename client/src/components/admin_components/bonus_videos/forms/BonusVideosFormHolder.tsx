import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/bonusVideoFormHolder.css";
// additional components //
import BonusVideoForm from "./BonusVideoForm";
import BonusVideoPreview from "../bonus_videos_preview/BonusVideoPreview";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createBonusVideo, editBonusVideo } from "../actions/APIBonusVideoActions";
import { closeBonusVideoForm, openBonusVideoForm } from "../actions/UIBonusVideoActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// type definitions //
import { BonusVideoData, FormState } from "../type_definitions/bonusVideoTypes"
import { RouteComponentProps } from "react-router-dom";

interface Props { }

const BonusVideoFormHolder: React.FC<Props> = ({ }): JSX.Element => {
  const { state, dispatch} = useContext(Store);
  // data for component //
  const { currentBonusVideoData, bonusVideoFormOpen } = state.bonusVideoState;
  const { description, youTubeURL, vimeoURL, createdAt, editedAt } = currentBonusVideoData;

  // form toggle listener //
  const toggleBonusVidform = () => {
    bonusVideoFormOpen ? closeBonusVideoForm(dispatch) : openBonusVideoForm(dispatch);
  }
  // CREATE - EDIT methods //
  const handleCreateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoData: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    createBonusVideo(bonusVideoData, dispatch)
      .then((_) => {
        closeBonusVideoForm(dispatch);
      })
      .catch((_) => {
        // handle error show modal ? //
      });
  };
  const handleUpdateBonusVideo = ({ description, youTubeURL, vimeoURL }: FormState): void => {
    const bonusVideoParams: BonusVideoData = {
      description,
      youTubeURL,
      vimeoURL
    };

    editBonusVideo(currentBonusVideoData._id, bonusVideoParams, dispatch, state)
      .then((_) => {
          closeBonusVideoForm(dispatch)
      })
      .catch((_) => {
        // handle an error or show a modal ? //
      })
  };
  // lifecycle hooks //
  useEffect(() => {
    if (!bonusVideoFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ bonusVideoFormOpen ]);
  // component render //
  return (
    <div id="bonusVideoFormHolder">
      {
        checkSetValues<IBonusVideoData>(currentBonusVideoData) ?
          <React.Fragment>
            <Grid.Row id="bonusVideoFormDetails">
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="bonusVideoFormHolderYouTubeUrl">
                  <h3>Video youTube URL:</h3>
                  <p>{youTubeURL}</p>
                </div>
                <div className="bonusVideoFormHolderVimeoUrl">
                  <h3>Video vimeo URL:</h3>
                  <p>{vimeoURL}</p>
                </div>
                <div className="bonusVideoFormHolderDescription">
                  <h3>Video description:</h3>
                  <p>{description}</p>
                </div>
                <div className="bonusVideoFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
                <div className="bonusVideoFormPreview">
                  <BonusVideoPreview bonusVideo={currentBonusVideoData} />
                </div>
              </Grid.Column>
            </Grid.Row>
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  
            id="bonusVideoFormToggleBtn" 
            onClick={toggleBonusVidform} 
            content={ bonusVideoFormOpen ? "Close Form" : "Open Form"}
          />
          {
            bonusVideoFormOpen ? 
            <BonusVideoForm 
              description={description} 
              youTubeURL={youTubeURL}
              vimeoURL={vimeoURL}
              handleCreateBonusVideo={handleCreateBonusVideo}
              handleUpdateBonusVideo={handleUpdateBonusVideo}
            /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

export default BonusVideoFormHolder;

