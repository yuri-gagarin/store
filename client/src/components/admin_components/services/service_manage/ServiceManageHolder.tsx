import React, { useContext, useEffect, useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import ServiceFormHolder from "../forms/ServiceFormHolder";
import ServiceCard from "./ServiceCard";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
// actions and state //
import { getAllServices } from "../actions/APIServiceActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";

interface Props extends RouteComponentProps {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
};

const ServiceManageHolder: React.FC<Props> = ({ state, dispatch, history }): JSX.Element => {
  const { loadedServices } = state.serviceState;
  // local state //
  const [ pageLoaded, setPageLoaded ] = useState<boolean>(false);
  const match = useRouteMatch("/admin/home/my_services/manage");

  const handleBack = () => {
    history.goBack();
  };
  useEffect(() => {
    getAllServices(dispatch)
      .then((success) => {
        if (success) {
          setPageLoaded(true);
        }
      });
  }, []); 

  if (pageLoaded) {
    return (
      <Grid padded stackable columns={2}>
        <Route path={match?.url + "/edit"}> 
          <Grid.Row>
            <Grid.Column computer={12} tablet={6} mobile={16}>
              <h3>Editing Service: { state.serviceState.currentServiceData.name }</h3>
              <Button inverted color="green" content="Back" onClick={handleBack}></Button>
            </Grid.Column>
          </Grid.Row>
          <ServiceFormHolder state={state} dispatch={dispatch} />
        </Route>
        <Route exact path={match?.url}>
          <Grid.Row>
            <Grid.Column computer={12} tablet={8} mobile={16}>
            {
              loadedServices.map((service) => {
                return (
                  <ServiceCard 
                    key={service._id}
                    service={service}
                    imageCount={service.images.length}
                    state={state}
                    dispatch={dispatch}
                  />
                );
              })
            }
            </Grid.Column>
            <Grid.Column computer={4} tablet={8} mobile={16}>
              
            </Grid.Column>
          </Grid.Row>
        </Route>
      </Grid>
    );
  } else {
    return (
      <LoadingScreen />
    );
  }
};

export default withRouter(ServiceManageHolder);