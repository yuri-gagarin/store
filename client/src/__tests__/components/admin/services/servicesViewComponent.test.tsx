import React, { ReactElement } from "react";
import { Menu } from "semantic-ui-react";
// testing imports //
import { mount, shallow, ReactWrapper, ShallowWrapper } from "enzyme";
// react router //
import { MemoryRouter as Router, Switch } from "react-router-dom";
import { AdminServiceRoutes } from "../../../../routes/adminRoutes";
// component imports //
import ServiceViewComponent from "../../../../components/admin_components/services/ServiceView";
import AdminServiceMenu from "../../../../components/admin_components/menus/AdminServiceMenu";
import { ServicePreviewHolder } from "../../../../components/admin_components/services/service_preview/ServicePreviewHolder";
import { ServiceFormHolder } from "../../../../components/admin_components/services/forms/ServiceFormHolder";
import { ServiceManageHolder } from "../../../../components/admin_components/services/service_manage/ServiceManageHolder";
// additional dependencies //


describe("ServiceView Component render tests", () => {
  // TEST ServiceView Component render //
  describe("ServiceView Component default render test", () => {
    let component: ShallowWrapper;

    beforeAll(() => {
      component = shallow(
        <ServiceViewComponent />
      );
    });

    it("Should Properly render Services View", () => {
      expect(component).toMatchSnapshot();
    });
    it("Should render conditional routes", () => {
      expect(component.find(Switch)).toHaveLength(1);
    });
    it("Should have Render three children route componets", () => {
      const switchComponent = component.find(Switch);
      expect(switchComponent.props().children).toHaveLength(3);
    });
    it("Should have proper client routes and conditionally render correct Components", () => {
      const switchComponent = component.find(Switch);
      type RouteMap = {
        [key: string]: string
      }
      let map: RouteMap = {};

      const routeComponents = switchComponent.props().children as ReactElement[];
      for (const component of routeComponents) {
        if (component.props.children[1].type.WrappedComponent) {
          // to check if a component is wrapped in a WithRouter() function //
          map[component.props.path as string] = component.props.children[1].type.WrappedComponent;
        } else {
          map[component.props.path as string] = component.props.children[1].type;
        }
      }
      expect(map[AdminServiceRoutes.VIEW_ALL_ROUTE]).toBe(ServicePreviewHolder);
      expect(map[AdminServiceRoutes.CREATE_ROUTE]).toBe(ServiceFormHolder);
      expect(map[AdminServiceRoutes.MANAGE_ROUTE]).toBe(ServiceManageHolder);
    });
  });
  // END ServicesViewComponent default render tests //
  // TEST AdminServiceMenu within component and Link button tets //
  describe("ServiceView Component button actions", () => {
    let component: ReactWrapper;

    beforeAll(() => {
      component = mount(
        <Router keyLength={0} initialEntries={ [AdminServiceRoutes.HOME_ROUTE] }>
          <ServiceViewComponent />
        </Router>
      );
    });

    describe("AdminServiceMenu", () => {

      it("Should render AdminServiceMenu component", () => {
        expect(component.find(AdminServiceMenu)).toHaveLength(1);
      });
      it("Should have 3 main navigation links", () => {
        const wrapper = component.find(AdminServiceMenu);
        const links = wrapper.find(Menu.Item)
        expect(links.length).toEqual(4);
      });
      it("'View All Services' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminServiceMenu).find(Menu.Item).at(0);
        viewAllLink.simulate("click");
        component.update();
        // assert updated component //
        expect(component.find(ServicePreviewHolder).length).toEqual(1);
        expect(component.find(ServiceFormHolder).length).toEqual(0);
        expect(component.find(ServiceManageHolder).length).toEqual(0);
      });
      it("'Create Service' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminServiceMenu).find(Menu.Item).at(1);
        viewAllLink.simulate("click");
        // assert updated component //
        component.update();
        expect(component.find(ServicePreviewHolder).length).toEqual(0);
        expect(component.find(ServiceFormHolder).length).toEqual(1);
        expect(component.find(ServiceManageHolder).length).toEqual(0);
      });
      it("'Manage Services' link should properly function", () => {
        window.scrollTo = jest.fn();
        const viewAllLink = component.find(AdminServiceMenu).find(Menu.Item).at(2);
        viewAllLink.simulate("click");
        component.update();
        // assert updated component //
        expect(component.find(ServicePreviewHolder).length).toEqual(0);
        expect(component.find(ServiceFormHolder).length).toEqual(0);
        expect(component.find(ServiceManageHolder).length).toEqual(1);
      });

    });
  });
  // END AdminServiceMenu link tets //

});