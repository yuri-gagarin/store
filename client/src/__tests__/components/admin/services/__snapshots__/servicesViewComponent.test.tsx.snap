// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceView Component render tests ServiceView Component default render test Should Properly render Services View 1`] = `
<div
  id="adminServiceViewHolder"
>
  <withRouter(AdminServiceMenu)
    dispatch={[Function]}
  />
  <Switch>
    <Route
      path="/admin/home/my_services/all"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(ServicePreviewHolder) />
    </Route>
    <Route
      path="/admin/home/my_services/create"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(ServiceFormHolder) />
    </Route>
    <Route
      path="/admin/home/my_services/manage"
    >
      <Component
        height="100px"
        width="100%"
      />
      <withRouter(ServiceManageHolder) />
    </Route>
  </Switch>
</div>
`;
